game/hud/src/services/session/layout.ts

```
@@ -109,18 +121,80 @@ const initialState = {
   widgets: defaultWidgets(),
 }
 
-let getInitialState = () => {
-  var storedState = JSON.parse(localStorage.getItem(localStorageKey)) as LayoutState;
+interface Size {
+  width: number;
+  height: number;
+}
+
+const px2pcnt = (pos: Position, screen: Size) : Position => {
+  // converts pixel co-ordinates into % of display size
+  return {
+    x: pos.x / screen.width,
+    y: pos.y / screen.height,
+    width: pos.width,           // don't scale width/height
+    height: pos.height,
+    scale: 1
+  };
+};
+
+const pcnt2px = (pos: Position, screen: Size) : Position => {
+  // converts % co-ordinates into pixels for current display
+  return {
+    x: screen.width * pos.x,
+    y: screen.height * pos.y,
+    width: pos.width,           // don't scale width/height
+    height: pos.height,
+    scale: 1
```

This function can make no assumptions about who/what is calling it. It should validate and fail in an expected manner if the required input is not provided. All functions should validate all objects before referencing members.